src/Microsoft.AspNetCore.Razor.Language/StreamSourceDocument.cs

```diff
@@ -0,0 +1,117 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal class StreamSourceDocument : RazorSourceDocument
+    {
+        // Internal for testing
+        internal readonly RazorSourceDocument _innerSourceDocument;
+
+        private readonly byte[] _streamBytes;
+        private byte[] _checksum;
+
+        public StreamSourceDocument(Stream stream, Encoding encoding, string fileName)
+        {
+            if (stream == null)
+            {
+                throw new ArgumentNullException(nameof(stream));
+            }
+
+            _streamBytes = ReadStreamBytes(stream);
+            _innerSourceDocument = CreateInnerSourceDocument(stream, encoding, fileName);
```

Comment:
> Rather than caching the bytes, should we just compute the hash as we read them?

Thought about doing this but saw it wouldn't change the `LargeTextSourceDocument` case (it needs to do a pre-pass over the stream to determine bytes). I'll actually code it out and see what it looks like.

> It just seems a little silly to keep the bytes around unless we have to.

Was trying to be lazy about the checksum calculation; with how the world currently works we always evaluate it anyways so I agree with you.

> Note btw that Roslyn specifically passes the encoding so that the exact bytes are round-trippable from the chars.

Hmmm, does getting the character representation of the BOM not result in issues when round tripping?